topics/dev/tutorials/debugging/tutorial.md

```diff
@@ -0,0 +1,730 @@
+time_estimation: "4h"
+key_points:
+  - "TODO Unit Tests"
+  - "TODO API Tests"
+  - "TODO Client Linting Failures"
+  - "TODO Client Tests"
+  - "TODO Selenium Tests"
+  - "TODO Runtime Errors"
+contributors:
+  - assuntad23
+  - ic4f
+  - jmchilton
+requirements:
+  - type: none
+    title: "Familiarity with basic Git commands"
+  - type: none
+    title: "Basic knowledge of Python and JavaScript"
+  - type: none
+    title: "Mac OS or Linux that can run Galaxy & your favorite IDE or editor"
+  - type: "internal"
+    topic_name: dev
+    tutorials:
+      - architecture
+subtopic: core
+---
+
+# Introduction
+{:.no_toc}
+
+
+In this tutorial we will demonstrate how to find and fix common types of bugs you may encounter as a contributor to Galaxy. We will step you through the process of finding and fixing a bug - from locating specific errors in the logs of Galaxy's GitHub workflows, to identifying their cause, developing a solution and committing your edits.
+
+With the skills from this tutorial, it is our hope that you will feel more prepared to develop solutions for Galaxy and more confidently navigate any obstacles along the way.
+
+> ### Agenda
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+# Local development environment setup
+
+For almost every type of error you encounter, there will be a relatively similar process:
+
+1. Locate failing test or error on GitHub workflows
+2. Run failing test locally
+3. Identify issue causing failure
+4. Develop a solution
+5. Ensure test passes locally
+6. Commit and push fixed branch to local fork
+
+The process for dealing with runtime errors will be somewhat different, since there will be no failed test to review initially. However, in either case, the first step is to setup your local development environment.
+
+## Contributing to Galaxy
+
+To contribute to galaxy, a GitHub account is required. Changes are proposed via a [pull request](https://docs.github.com/en/github/collaborating-with-pull-requests). This allows the project maintainers to review the changes and suggest improvements.
+
+The general steps are as follows:
+1. Fork the Galaxy repository
+2. Clone your fork
+3. Make changes in a new branch
+4. Commit your changes, push branch to your fork
+5. Open a pull request for this branch in the upstream Galaxy repository
+
+In this tutorial, instead of cloning Galaxy's default branch, we will clone a branch that has been created for this tutorial and contains several bugs that you'll need to identify and fix. Also, we won't be opening a pull request in the upstream Galaxy repository.
+
+## Getting started: Pulling the buggy branch
+
+Good news! All of the failing tests are on the same Galaxy branch! That means you only need to download and work off of one branch. The downside is that this is one buggy branch with issues all over the place. We have no doubts that you can solve them though!
+
+> ### {% icon hands_on %} Hands-on: Setup your local Galaxy instance
```

Comment:
to be honest, I am not sure if we need this section. all this can be created with `sh run.sh`
Do we really want them to configure virtualenv themselves?

Reply:
It's much faster than `run.sh`, and we don't need `run.sh` to run tests in the first part of the tutorial. I think that'll get them going faster; besides, that may be yet another useful thing to learn - it might take the mystery out of .venv for some people, I think.

Reply:
maybe (just to skip the installation part), we can show them how to run 1 random test with `run_tests.sh`? It should create .venv as well

Reply:
We do ask them to use `run _tests.sh` for the api and selenium sections. Also, I've added this line that says that run_tests.sh will activate the venv (not create, since we ask them to create it at the beginning of the tutorial, so they already have it):
https://github.com/galaxyproject/training-material/pull/2614/files#diff-03c321ac89fe848061d63bb53e4d75eec9a2f914b4b13efa2f86afae057572a4R238